### PR TITLE
ENH: Vendor meson/meson-python for multi-target build support

### PR DESCRIPTION
**Pull request by rgommers**
The purpose of this PR is to enable merging the SIMD support for the Meson builds in gh-23096. That PR relies on https://github.com/mesonbuild/meson/pull/11307, which is a significant new feature for Meson. That is multi-target support, allowing us to recompile the same sources multiple times, with compile flags set for different sets of SIMD instructions. Given that that PR is not going to get merged into Meson in time, we'll want to vendor it in order to be able to use it now.

This PR vendors the following:
- `meson` as of [commit 9d88d0d5c](https://github.com/mesonbuild/meson/commit/9d88d0d5c) (1.2.1-dev, last change 18 July 2023), plus the `feature` module from https://github.com/mesonbuild/meson/pull/11307
- `meson-python` as of [commit 2e99c7b5bc](https://github.com/mesonbuild/meson-python/commit/2e99c7b5bc) (current HEAD, 0.14.0-dev0, last change 26 July 2023)
- [`spin/cmds/meson.py`](https://github.com/scientific-python/spin/blob/main/spin/cmds/meson.py)

And then does the required plumbing in order to use the vendored code for building.

I have tested that the combination of this PR and gh-23096 works together, and that the sdist/wheels produced look the way they should. The plan is to merge this PR first, and then rebase gh-23096 on top (Cc @seiko2plus).

I chose vendoring everything over packaging up a separate package (something like `meson-numpy`), because that'd be a real pain for distributions, they'd have to package that in all of their distros as a new package in other to be able to build numpy 1.26.0.

Licensing-wise: Meson is Apache 2.0 licensed, but because we don't install it as part of a built numpy package, I don't think that's a problem.

@eli-schwartz in case you have any insights or concerns here regarding the approach, or things you'd like to see in terms of validation for https://github.com/mesonbuild/meson/pull/11307, please let us know.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,8 +1,18 @@
...
 build-backend = "mesonpy"
+backend-path = ['./vendored-meson/meson-python']
 requires = [
     "Cython>=3.0",
-    "meson-python>=0.13.1",
+    # All dependencies of the vendored meson-python (except for meson, because
+    # we've got that vendored too - that's the point of this exercise).
+    'pyproject-metadata >= 0.7.1',
+    'tomli >= 1.0.0; python_version < "3.11"',
+    'setuptools >= 60.0; python_version >= "3.12"',
```
Comment (charris) [inline pyproject.toml:10]: A more recent version of setuptools is also needed for Python 3.11 due to deprecations. I'm running 68.0.0 at the moment. There may be a sweet spot in between, but I don't know what it is.

**Comment by eli-schwartz** [comment]
> @eli-schwartz in case you have any insights or concerns here regarding the approach, or things you'd like to see in terms of validation for [mesonbuild/meson#11307](https://github.com/mesonbuild/meson/pull/11307), please let us know.

I think this is quite understandable in order to move forward with numpy, it's a very friendly fork indeed. However, it seems a bit unfortunate to vendor meson-python as well, just to set the meson_cli path (I think?). I wonder if it would be possible to instead add your own tiny in-tree build-backend that:
- manipulates `$PATH` to include `./vendored-meson/entrypoint`
- imports * from mesonpy

And stick a wrapper in that directory which invokes the meson.py script properly.

It may also make sense to include meson as a submodule -- this would probably be friendlier for `git log`. It shouldn't be hard to rebase that feature branch if you need fixes in between `1.2.0` and `master`. If you use a wrapper build backend around mesonpy, you could even try `git submodule update --init` automatically.

File: pyproject.toml (hunk 1 above)
Comment (charris) [reply pyproject.toml:10]: Looks like 62.3.0 is the first version to have the fixes for 3.11, might go later just to be safe.

**Comment by mattip** [comment]
> It may also make sense to include meson as a submodule -- this would probably be friendlier for git log

+1

**Comment by rgommers** [comment]
> I wonder if it would be possible to instead add your own tiny in-tree build-backend that:

Good idea, thanks. I'll give that a try this morning. 

>  it seems a bit unfortunate to vendor meson-python as well, just to set the meson_cli path (I think?)

I also edited the meson-python dependencies to remove `meson`, but installing that and then not using it is not really a problem.

> It may also make sense to include meson as a submodule

I think the main issue is keeping that working permanently. The PR branch is likely to be rebased multiple times, so we can't rely on it. So we'd have to create a repo at https://github.com/numpy/meson with branches or tags to point to, and keep it there forever. I hope this is a short-lived vendoring, and it can go away towards the end of the year (please correct me if you think that that is unlikely to be the case). 

We can go this way if other maintainers prefer this, but I think I personally have a preference for not having another repo around permanently.

**Comment by mattip** [comment]
> We can go this way if other maintainers prefer this, but I think I personally have a preference for not having another repo around permanently.

I think it will be easier to reason about upstream changes if we manage a fork. Meson seems to still be under active development (over [12 PRs merged in 3 days](https://github.com/mesonbuild/meson/pulls?q=is%3Apr+sort%3Aupdated-desc+is%3Aclosed), and while we may not need many of those changes, managing that churn will be easier if it is a separate repo. Once our changes are merged upstream, we can archive the repo.

**Comment by rgommers** [comment]
Okay, thanks Matti. I'll create a new `numpy/meson` repo then and will update this PR.

**Comment by rgommers** [comment]
The approach in gh-24379 works and seems preferable, so closing this PR.

**Pull request closed without merging**